File: python/day09.py
# DAY09
import math

OFFS = [(0, 1), (0, -1), (1, 0), (-1, 0)]

def find_minima(board):
    res = []
    for j, row in enumerate(board):
        for i, h in enumerate(row):
            is_min = True
            for dx, dy in OFFS:
                x, y = i + dx, j + dy
                if x >= 0 and x < len(row) and y >= 0 and y < len(board) and board[y][x] <= h:
                    is_min = False
                    break
            if is_min:
                res.append((i, j))
    return res

def fill_cell(board, bidx, pos, basin_idx):
    x, y = pos
    if x < 0 or x >= len(board[0]) or y < 0 or y >= len(board):
        return
    b = board[y][x]
    if b == 9:
        bidx[y][x] = 0
        return

    if bidx[y][x] >= 0:
        return
    bidx[y][x] = basin_idx
    h = board[y][x]
    for dx, dy in OFFS:
        x1, y1 = x + dx, y + dy
        fill_cell(board, bidx, (x1, y1), basin_idx)

def find_basins(board):
    w, h = len(board[0]), len(board)
    bidx = [[-1]*w for i in range(h)]
    num_basins = 0
    for j in range(h):
        for i in range(w):
            if bidx[j][i] == -1:
                fill_cell(board, bidx, (i, j), num_basins + 1)
                num_basins += 1
    return bidx

def solution():
    board = [list(map(int, list(x.strip()))) for x in open('../data/09.txt').readlines()]

    mins = find_minima(board)
    res1 = sum(map(lambda p: board[p[1]][p[0]] + 1, mins))

    basins = find_basins(board)
    max_basin = max(map(max, basins))

    basin_sizes = []
    for bidx in range(1, max_basin + 1):
        size = 0
        w, h = len(board[0]), len(board)
        for j in range(h):
            for i in range(w):
                if basins[j][i] == bidx:
                    size += 1
        basin_sizes.append(size)

    res2 = 1
    for x in sorted(basin_sizes, reverse=True)[:3]:
        res2 *= x
    print(f"Answer 1: {res1}\nAnswer 2: {res2}")

File: rust/src/day09.rs
use crate::common;

const OFFS: [(i32, i32); 4] = [(0, 1), (0, -1), (1, 0), (-1, 0)];

fn find_minima(board: &Vec<Vec<u8>>) -> Vec<(usize, usize)> {
    let mut res = vec![];
    let bh = board.len() as i32;
    for (j, row) in board.iter().enumerate() {
        for (i, h) in row.iter().enumerate() {
            let bw = row.len() as i32;
            let is_min = OFFS.iter().all(|(dx, dy)| {
                let x = i as i32 + dx;
                let y = j as i32 + dy;
                x < 0 || x >= bw || y < 0 || y >= bh || board[y as usize][x as usize] > *h
            });
            if is_min {
                res.push((i, j));
            }
        }
    }
    res
}

fn fill_cell(board: &Vec<Vec<u8>>, bidx: &mut Vec<Vec<i32>>, pos: (usize, usize), basin_idx: i32) {
    let (x, y) = pos;
    if x >= board[0].len() || y >= board.len() {
        return;
    }
    let b = board[y][x];
    if b == 9 {
        bidx[y][x] = 0;
        return;
    }
    if bidx[y][x] >= 0 {
        return;
    }
    bidx[y][x] = basin_idx;
    for (dx, dy) in OFFS {
        let (x1, y1) = ((x as i32 + dx) as usize, (y as i32 + dy) as usize);
        fill_cell(&board, bidx, (x1, y1), basin_idx);
    }
}

fn find_basins(board: &Vec<Vec<u8>>) -> Vec<Vec<i32>> {
    let w = board[0].len();
    let h = board.len();
    let mut bidx = vec![vec![-1; w]; h];
    let mut nbasins = 0;
    for j in 0..h {
        for i in 0..w {
            if bidx[j][i] == -1 {
                fill_cell(&board, &mut bidx, (i, j), nbasins + 1);
                nbasins += 1;
            }
        }
    }
    bidx
}

pub(crate) fn solution() {
    let lines = common::read_lines(&common::data_file(9)).unwrap();
    let board: Vec<Vec<u8>> = lines
        .iter()
        .map(|x| x.as_bytes().iter().map(|c| c - '0' as u8).collect())
        .collect();
    let mins = find_minima(&board);
    let res1: usize = mins.iter().map(|(x, y)| board[*y][*x] as usize + 1).sum();

    let basins = find_basins(&board);
    let max_basin = basins
        .iter()
        .map(|x| x.iter().max().unwrap())
        .max()
        .unwrap();

    let mut basin_sizes = vec![];
    let w = board[0].len();
    let h = board.len();
    for b in 1..max_basin + 1 {
        let mut size = 0;
        for j in 0..h {
            for i in 0..w {
                if basins[j][i] == b {
                    size += 1;
                }
            }
        }
        basin_sizes.push(size);
    }
    basin_sizes.sort();
    basin_sizes.reverse();
    let res2: usize = basin_sizes[0..3].iter().product();
    println!("Answer 1: {:?}\nAnswer 2: {:?}", res1, res2);
}
